### PR TITLE
Region Detection Refactoring

### PR DESCRIPTION
**Pull request by andreitudorp**

### CODE REVIEW

File: utils/regiondetection.py [hunk 1]
```diff
@@ -1,243 +1,314 @@
...
 import random
 import math
 
-# Setup logging
+# Set up logging
 from utils.customlogger import CustomLogger
 main_logger = CustomLogger().main_logger
 
+# Get the number of unique colours and the percentage of the primary colour in the given image
+def _count_colours(image : cv2.typing.MatLike):
```
Comment (TPGamesNL) [inline utils/regiondetection.py:15]: if you want to document functions, use python doc strings

File: utils/regiondetection.py [hunk 2]
```diff
@@ -1,243 +1,314 @@
...
     contours, hier = cv2.findContours(img, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
-
+    
     main_logger.debug("Storing valid contours")
-    roi = []
-    count = 0
+    
+    regions = []
+
     if len(contours) != 0:
```
Comment (TPGamesNL) [inline utils/regiondetection.py:89]: unnecessary check

File: utils/regiondetection.py [hunk 3]
```diff
@@ -1,243 +1,314 @@
...
-                # (see paper by Evdoxios Baratis and Euripides G.M. Petrakis why this is)
+            # Always true - region constraints could be applied here
+            
+            unique_colors_count, pct = _count_colours(region)
+            # also get a greyscale version of the region for the other attributes
+            # (see paper by Evdoxios Baratis and Euripides G.M. Petrakis why this is)
 
-                if (r.size == 0):
+            if (region.size == 0):
                     continue
```
Comment (TPGamesNL) [inline utils/regiondetection.py:104]: indentation

File: utils/regiondetection.py [hunk 4]
```diff
@@ -1,243 +1,314 @@
...
+            # TODO: Add subregion drawing
+            #if subregiondraw:
+            #     cv2.imwrite(f"{highlightname}.subregion.{idx}.png", region[0])
+            
+        cv2.imwrite(os.path.join(os.path.dirname(os.path.realpath(img_path)), f"{highlight_name}.png"), draw_img)
+
+# TODO: Add drawing capabilities for recursive and try to split the function into smaller parts
+def _find_regions(image : cv2.typing.MatLike, image_path : str, draw : bool, highlight_name : str, invert = True):
+    
+    main_logger.debug("Obtaining grayscale version of image")
```
Comment (TPGamesNL) [inline utils/regiondetection.py:64]: I think it'd be nice to seperate this image pre-processing from the region detection post-processing

File: utils/regiondetection.py [hunk 5]
```diff
@@ -1,243 +1,314 @@
...
+        
+    return regions_of_interest
+
+FLAG_NO_DRAW = 0
+"""Tells the function to **NOT** draw any debugging images.""" 
+
+FLAG_DRAW = 1
+"""Tells the function to draw the regions on the original image.""" 
 
+FLAG_DRAW_RECUSRIVE = 2
```
Comment (TPGamesNL) [inline utils/regiondetection.py:167]: Would be better to use Enum I think

File: utils/regiondetection.py [hunk 6]
```diff
@@ -1,243 +1,314 @@
...
+# Deprecated
+def findregions(imgpath, draw=True, recursivedraw=False, subregiondraw=False, highlightname="Highlight"):
+    """
+    This function is deprecated. Use find_regions() instead.
+    ------
+    """
+    
+    return find_regions(imgpath, draw_flag = FLAG_DRAW_RECUSRIVE, highlight_name = highlightname)
+
+# Deprecated
```
Comment (TPGamesNL) [inline utils/regiondetection.py:210]: for this and above: if deprecated and unused, you can remote it

File: utils/regiondetection.py (hunk 4 above)
Comment (andreitudorp) [reply utils/regiondetection.py:64]: Agree

File: utils/regiondetection.py (hunk 5 above)
Comment (andreitudorp) [reply utils/regiondetection.py:167]: Will look into that

File: utils/regiondetection.py [hunk 7]
```diff
@@ -1,243 +1,314 @@
...
+    """
+    Finds all the regions of interest in the image and returns the data of the image.
+    
+    Args:
+        img_path (str): The path (location) of the image.
+        draw_flag (int, optional): The flag to tell the function to draw the regions. Defaults to FLAG_DRAW.
+        highlight_name (str, optional): The name of the file to save the highlighted image. Defaults to "Highlight".
+    """
+    
+    if(draw_flag < FLAG_NO_DRAW or draw_flag > FLAG_DRAW_RECUSRIVE):
```
Comment (TPGamesNL) [inline utils/regiondetection.py:180]:
```suggestion
    if draw_flag < FLAG_NO_DRAW or draw_flag > FLAG_DRAW_RECUSRIVE:
```

**Review by TPGamesNL — APPROVED** [review]

**Pull request merged**